lib/Serialization/DeserializeSIL.cpp

```diff
@@ -607,6 +608,36 @@ SILDeserializer::readSILFunctionChecked(DeclID FID, SILFunction *existingFn,
         SILMod, requirements, exported != 0, specializationKind));
   }
 
+  // Read and instantiate the differentiable attributes.
+  while (numReverseDifferentiableAttrs--) {
+    auto next = SILCursor.advance(AF_DontPopBlockAtEnd);
+    assert(next.Kind == llvm::BitstreamEntry::Record);
+
+    scratch.clear();
+    kind = SILCursor.readRecord(next.ID, scratch);
+    assert(kind == SIL_REVERSE_DIFFERENTIABLE_ATTR &&
+           "Missing reverse differentiable attribute");
+
+    uint64_t primalNameId;
+    uint64_t adjointNameId;
+    uint64_t source;
+    ArrayRef<uint64_t> parameters;
+    SILReverseDifferentiableAttrLayout::readRecord(scratch, primalNameId,
+                                                   adjointNameId, source,
+                                                   parameters);
+
+    StringRef primalName = MF->getIdentifier(primalNameId).str();
```

Comment:
Both primal name and adjoint name can be null. Serialization/deserialization should take that into account.

If it's hard to fix, feel free to leave a todo here and in serialization.

Reply:
Ah, thanks, I wasn't aware!

I went and checked, and all the functions that I call in serialization/deserialization have logic for propagating nulls correctly and serializing/deserializing them to `0`. So it should already work. Also, most (all?) of the functions in the stdlib have null primals, so that is also evidence that it works.

Reply:
Stdlib functions shouldn't have null primals. When primal is not specified in a `@differentiable` attribute, the attribute will be lowered to a `[reverse_differentiable]` attribute where the primal is the original function itself.

Reply:
And, this SGTM!